python-package/lightgbm/basic.py

```diff
@@ -1640,6 +1640,11 @@ def add_features_from(self, other):
         if self.handle is None or other.handle is None:
             raise ValueError('Both source and target Datasets must be constructed before adding features')
         _safe_call(_LIB.LGBM_DatasetAddFeaturesFrom(self.handle, other.handle))
+        if other.data is None:
+            self.data = None
+        if self.data is not None:
+            # FIXME: concat two dataset
+            self.data = [self.data, other.data]
```

Comment:
@StrikerRUS here may need to concat two data by col.

Reply:
@guolinke Should we consider all possible combinations of supported data sources?
https://github.com/microsoft/LightGBM/blob/1c1a276519439bdb10a91855bc8af8801cd819d5/python-package/lightgbm/basic.py#L1532-L1541

Reply:
maybe we can convert the type of `other.data` to the type of `self.data` ?

Reply:
OK, I'll try to do something.